src/main/scala/io/scalac/lab/api/endpoints4s/ApartmentsEndpointsServer.scala

```diff
@@ -33,18 +31,16 @@ class ApartmentsEndpointsServer(storage: ApartmentsStorage, security: SecuritySe
   val deleteApartmentRoute: Route = deleteApartment.implementedByAsync { case (id, _) => storage.delete(id) }
 
   override def authenticatedRequest[A, B](request: Directive1[A])(implicit tupler: Tupler.Aux[A, ApiKey, B]): Directive1[B] = {
-    optionalHeaderValueByName("api-key").flatMap { authToken: Option[String] =>
-      onComplete(security.authenticate(authToken)).flatMap {
-        case Success(Right(apiKey)) => request.map(r => tupler(r, apiKey))
-        case _                      => complete((Unauthorized, authToken.fold("Missing API key")(_ => "Incorrect API key")))
+    request.flatMap { entity =>
```

Comment:
What is an advantage of making `request.flatMap{ entity => ... }`, instead of mapping the request later on?

Reply:
IIRC, the previous version was returning `Unauthorized` responses even for “non-authenticated” requests (ie, public endpoints).
It boils down to the way Akka-HTTP works when we combine directives with `~` (maybe you have already experienced that: you need to carefully `seal` some directives otherwise all your endpoints may be affected by one directive).
This quick fix makes sure the `request` directive did succeed before we even try to check the `Authorization` header. This is not perfect though because a malicious (unauthenticated) user could flood your server with high volume of data and you would happily accept its request entities before eventually rejecting them when you finally check the `Authorization` header.
A better solution would be to still perform the authentication check before decoding the request entity, but I think I couldn’t do this without changing the signature of `authenticatedRequest`, which I didn’t want to do in this PR.